### PR TITLE
Partial request assertion

### PR DESCRIPTION
**Pull request by squadgazzz**
### Background

https://github.com/gnosis/solvers/pull/37 introduced a random value in the POST request. This change presents challenges for our existing testing framework, as the random value prevents the use of static mock responses. To address this issue and ensure that we can still perform robust and accurate tests, the PR implements a new mechanism for partial assertions of requests.

### Solution

This pull request's key enhancement is the introduction of macros for partially asserting the contents of POST requests. This method allows us to validate the request structure and static fields while excluding the dynamically generated random values from the assertion checks.

### Changes

- [x] Start using 2-min deadlines for requests
- [x] Exclude the deadline from the POST request assertions in the tests

### CODE REVIEW

File: src/tests/mock/http.rs [hunk 1]
```diff
@@ -221,6 +227,43 @@ fn get(state: State, path: Option<String>, query: Option<String>) -> serde_json:
...
+/// excluding the ignored paths.
+///
+/// # Examples
+///
+/// ```
+/// let actual = json!({"user": {"id": 1, "name": "Alice", "email": "alice@example.com"}});
+/// let expected = json!({"user": {"id": 1, "name": "Alice", "email": "bob@example.com"}});
+/// assert_json_matches!(actual, expected, ["user.email"]);
+/// ```
+macro_rules! assert_json_matches {
```
Comment (MartinquaXD) [inline src/tests/mock/http.rs:257]: Is there a big selling point why this needs to be a macro?

File: src/tests/mock/http.rs [hunk 2]
```diff
@@ -255,3 +305,155 @@ fn full_path(path: Option<String>, query: Option<String>) -> String {
...
+fn parse_field_paths(paths: &[&str]) -> HashSet<Vec<String>> {
+    paths
+        .iter()
+        .map(|path| path.split('.').map(String::from).collect())
+        .collect()
+}
+
+/// Recursively compares two JSON values, excluding specified paths, and returns
+/// detailed errors using anyhow.
+fn json_matches_excluding(
```
Comment (MartinquaXD) [inline src/tests/mock/http.rs:319]: This seems quite complex. Wouldn't it be easier to do the following?
1. clone the actual value
2. delete the ignored paths from it
3. use strict equality again

That way we don't need recursion and the code would probably be self explanatory.

File: src/tests/mock/http.rs [hunk 3]
```diff
@@ -80,6 +82,10 @@ pub enum Expectation {
 pub enum RequestBody {
     /// The received `[RequestBody]` has to match the provided value exactly.
     Exact(serde_json::Value),
+    /// The received `[RequestBody]` has to match the provided value partially
+    /// excluding the specified paths which are represented as dot-separated
+    /// strings.
+    Partial(serde_json::Value, Vec<String>),
```
Comment (MartinquaXD) [inline src/tests/mock/http.rs:88]: Let's go for simple changes until we actually need something more complex.
```suggestion
    Partial(serde_json::Value, Vec<&'static str>),
```

File: src/infra/dex/balancer/mod.rs [hunk 4]
```diff
@@ -80,6 +81,8 @@ impl Sor {
             self.chain_id,
             self.settlement,
             self.query_batch_swap,
+            // 2 minutes from now
+            (chrono::Utc::now().timestamp() + 120).to_u64(),
```
Comment (Mateo-mro) [inline src/infra/dex/balancer/mod.rs:85]: What about summing up the UTC time in seconds and then transforming it to timestamp? :thinking: it is clearer we are adding two minutes

File: src/tests/mock/http.rs [hunk 5]
```diff
@@ -255,3 +305,155 @@ fn full_path(path: Option<String>, query: Option<String>) -> String {
...
+        .collect()
+}
+
+/// Recursively compares two JSON values, excluding specified paths, and returns
+/// detailed errors using anyhow.
+fn json_matches_excluding(
+    actual: &serde_json::Value,
+    expected: &serde_json::Value,
+    exclude_paths: &HashSet<Vec<String>>,
+    current_path: &mut Vec<String>,
```
Comment (Mateo-mro) [inline src/tests/mock/http.rs:323]: Please can you add in the function docs what to expect to be modified in this field?

File: src/tests/mock/http.rs [hunk 6]
```diff
@@ -221,6 +227,43 @@ fn get(state: State, path: Option<String>, query: Option<String>) -> serde_json:
...
+/// ```
+/// let actual = json!({"user": {"id": 1, "name": "Alice", "email": "alice@example.com"}});
+/// let expected = json!({"user": {"id": 1, "name": "Alice", "email": "bob@example.com"}});
+/// assert_json_matches!(actual, expected, ["user.email"]);
+/// ```
+macro_rules! assert_json_matches {
+    ($actual:expr, $expected:expr, $exclude_paths:expr) => {{
+        // Encapsulate all operations inside a block to scope variables properly
+        let exclude_paths = parse_field_paths(&$exclude_paths);
+        let mut path = vec![];
```
Comment (Mateo-mro) [inline src/tests/mock/http.rs:261]: Defining an unused variable input just to make the function recursive doesn't seem to be the best practice. Why don't you make two functions? the wrapper (which defines this variable) and the one that calls itself?

Comment (squadgazzz) [reply src/tests/mock/http.rs:261]: This macro is actually the wrapper.

File: src/tests/mock/http.rs (hunk 1 above)
Comment (squadgazzz) [reply src/tests/mock/http.rs:257]: To alight with the current code https://github.com/gnosis/solvers/blob/cd4606bab5d7f4f02a654b49f7f1b523c7db6cd0/src/tests/mock/http.rs#L285 and also this produced meaningful error messages. And this macro could be used in other tests in the future.

File: src/tests/mock/http.rs (hunk 2 above)
Comment (squadgazzz) [reply src/tests/mock/http.rs:319]: I was a bit tired of seeing "right doesn't equal left" which requires manually finding the differences, which is annoying.

File: src/tests/mock/http.rs [hunk 7]
```diff
@@ -221,6 +227,43 @@ fn get(state: State, path: Option<String>, query: Option<String>) -> serde_json:
...
+        json_matches_excluding(&$actual, &$expected, &exclude_paths, &mut path)
+            .expect("JSON did not match with the exclusion of specified paths");
```
Comment (Mateo-mro) [inline src/tests/mock/http.rs:263]: You are squashing the error generated by the function with the expect text

Comment (squadgazzz) [reply src/tests/mock/http.rs:263]: Not really, here you can see how it works: https://github.com/gnosis/solvers/blob/d6b63fe9c383c4f4140f92baee6f9cfee162e5a7/src/tests/mock/http.rs#L434-L437

File: src/tests/mock/http.rs (hunk 6 above)
Comment (squadgazzz) [reply src/tests/mock/http.rs]: Added docs and created a helper function.

File: src/tests/mock/http.rs (hunk 5 above)
Comment (squadgazzz) [reply src/tests/mock/http.rs]: Added docs

File: src/tests/mock/http.rs [hunk 8]
```diff
@@ -255,3 +303,196 @@ fn full_path(path: Option<String>, query: Option<String>) -> String {
...
+        let json_b = json!({
+            "user": {
+                "id": 123,
+                "profile": {
+                    "name": "Alice",
+                    "timestamp": "2022-01-01T12:00:00Z"
+                }
+            }
+        });
+        assert_json_matches!(json_a, json_b, ["user.profile.timestamp"])
```
Comment (Mateo-mro) [inline src/tests/mock/http.rs:443]: Please, could you add a test with more than one element in the array? :pray:

Comment (squadgazzz) [reply src/tests/mock/http.rs]: Added

**Review by mateo-mro — APPROVED** [review]

File: src/tests/mock/http.rs (hunk 2 above)
Comment (MartinquaXD) [reply src/tests/mock/http.rs:317]: I think giving better error messages is a good reason. But if this is then intended as a more versatile building block I think we probably should put it into `shared` so we can also make use of it in the backend repo while not duplicating code. 🤔

Comment (squadgazzz) [reply src/tests/mock/http.rs:317]: https://github.com/cowprotocol/services/pull/2783

**Pull request closed without merging**